compiler/rustc_mir_transform/src/shim.rs

```diff
@@ -1051,12 +1016,17 @@ fn build_fn_ptr_addr_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, self_ty: Ty<'t
 fn build_construct_coroutine_by_move_shim<'tcx>(
     tcx: TyCtxt<'tcx>,
     coroutine_closure_def_id: DefId,
+    receiver_by_ref: bool,
 ) -> Body<'tcx> {
-    let self_ty = tcx.type_of(coroutine_closure_def_id).instantiate_identity();
+    let mut self_ty = tcx.type_of(coroutine_closure_def_id).instantiate_identity();
     let ty::CoroutineClosure(_, args) = *self_ty.kind() else {
         bug!();
     };
 
+    if receiver_by_ref {
+        self_ty = Ty::new_mut_ptr(tcx, self_ty);
```

Comment:
since it's raw pointers, it doesn't matter that it's mutable for `AsyncFn`? Or we need it to be mutable anyway, because we will keep polling the coroutine?

Reply:
It doesn't matter that it's mutable; it just needs to be ABI-compatible.

The self type here is a coroutine-closure, not a coroutine, and we never read from it because it never has any captures (because this is the `Fn`/`FnMut` implementation, not the `AsyncFn`/`AsyncFnMut` implementation, which is implemented *only if* the coroutine-closure has no captures).

I'll leave a comment for that.